lib/Extractor/KLEEBuilder.cpp

```diff
@@ -169,6 +172,21 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
                                                                                   klee::ConstantExpr::create(i, Width))));
       PowerTwoBitsMap[Origin] = PowerExpr;
     }
+    if (Origin->NumSignBits > 1) {
+      ref<Expr> SignBitsExpr = klee::ConstantExpr::create(1, Expr::Bool);
+      ref<Expr> Res = LShrExpr::create(Var, klee::ConstantExpr::create(Width-Origin->NumSignBits, Width));
+      ref<Expr> TestOnes = SubExpr::create(ShlExpr::create(klee::ConstantExpr::create(1, Width),
+                                           klee::ConstantExpr::create(Origin->NumSignBits, Width)),
+                                           klee::ConstantExpr::create(1, Width));
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     OrExpr::create(EqExpr::create(Res, TestOnes),
+                                                    EqExpr::create(Res, klee::ConstantExpr::create(0, Width))));
+      ref<Expr> MSB = ExtractExpr::create(Var, Width-1, Expr::Bool);
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     Expr::createImplies(EqExpr::create(MSB, klee::ConstantExpr::create(0, Expr::Bool)),
+                                                         EqExpr::create(ctlz(Var), klee::ConstantExpr::create(Origin->NumSignBits, Width))));
```

Comment:
I'm not sure this is correct -- it looks like you're asserting the the number of sign bits is equal to the value. This is too strong. The number of sign bits is a conservative estimate: it has to be at least that much, but might be more. You can's say that it's exactly a certain value.

I think there's an easier way to do what you are trying to do here: just assert that the highest N bits of the value are either all 0 or all 1.  

Or better yet, shift the value right (arithmetically) by width-#signbits and then assert that the resulting value is either 0 or all ones.